rewrite shim generator

onerandomusername opened the pull request:
- use argparse and pathlib
- format the imports with isort
- include the license with each file
- fix ext.tasks shim not being accurate
- fix a bug with discord.types.snowflake not being importable

onerandomusername committed:
feat: use argparse and pathlib, re-export tasks init to fix bug

feat: rewrite the majority of the shim maker

shiftinv commented on discord/opus.py:
```diff
@@ -19,16 +45,18 @@
     _load_default,
+    _OpusStruct,
     band_ctl,
+    c_float_ptr,
```
just so this won't be forgotten: these three aren't imported when `shim_maker.py` is run on Windows

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+def shim_folder(path: pathlib.Path, pypath: str, shim_path: pathlib.Path) -> List[pathlib.Path]:
+    modified = []
+    for fn in path.iterdir():
+        mod = fn.relative_to(path)
```
```suggestion
        shim_mod = shim_path / fn.relative_to(path)
```

And replace all `shim_path / mod` with `shim_mod` below, as well as `mod` with `fn`
makes it a bit less repetitive

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+            print("Created directory: ", (shim_path / mod))
+
+        if fn.name in ("__init__.py", "__main__.py", "py.typed") and "tasks" != fn.parent.name:
+            with open(path / mod, "r", encoding="utf-8") as fr:
+                try:
+                    with open(shim_path / mod, "r", encoding="utf-8") as fw:
+                        if fw.read() == fr.read():
+                            continue
+                except FileNotFoundError:
+                    pass
+                fr.seek(0)
+                print("Updating file: ", shim_path / mod)
+                with open(shim_path / mod, "w", encoding="utf-8") as fw:
+                    fw.write(fr.read())
+            modified.append(shim_path / mod)
```
could be simplified to
```suggestion
            data = fn.read_text(encoding="utf-8")
            if shim_mod.is_file() and data == shim_mod.read_text(encoding="utf-8"):
                continue
            print("Updating file: ", shim_mod)
            shim_mod.write_text(data, encoding="utf-8")
            modified.append(shim_mod)
```

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+            if docstring:
+                to_write += '"""\n' + docstring.strip() + '\n"""\n\n'
+            to_write += imports
+            try:
+                with open(shim_path / mod) as f:
+                    existing = f.read()
+            except FileNotFoundError:
+                existing = None
+            if existing != to_write:
+                print("Updating file: ", shim_path / mod)
+                with open(shim_path / mod, "w", encoding="utf-8") as f:
+                    f.write(to_write)
+
+                modified.append(shim_path / mod)
```
could also be simplified to
```suggestion

            if shim_mod.is_file():
                existing = shim_mod.read_text(encoding="utf-8")
            else:
                existing = None
            if existing != to_write:
                print("Updating file: ", shim_mod)
                shim_mod.write_text(to_write, encoding="utf-8")
                modified.append(shim_mod)
```

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+    return modified
+
+
+def shim(path: pathlib.Path, shim_path: pathlib.Path) -> None:
```
```suggestion
def shim(path: pathlib.Path, shim_path: pathlib.Path) -> List[pathlib.Path]:
```

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+    return shim_folder(path, path.stem, shim_path)
+
+
+def main():
```
```suggestion
def main() -> List[pathlib.Path]:
```

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+    try:
+        path = pathlib.Path(args.path)
+    except ValueError:
+        print(f"Invalid path to package to be shimmed: {args.path}")
+        return 2
```
```suggestion
        print(f"Invalid path to package to be shimmed: {args.path}", file=sys.stderr)
        sys.exit(2)
```

onerandomusername replied:
Won't implement this, the main method shouldn't exit.

shiftinv replied:
The reason for doing this was that `len(res)` below could otherwise fail, since `res` is `List[Path] | Literal[2, 1, 0]`

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+    try:
+        shim_dir = pathlib.Path(args.shim)
+    except ValueError:
+        print(f"Invalid path to new package: {args.shim}")
+        return 2
```
```suggestion
        print(f"Invalid path to new package: {args.shim}", file=sys.stderr)
        sys.exit(2)
```

onerandomusername replied:
Won't implement this, the main method shouldn't exit.

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
-    os.makedirs("discord", exist_ok=True)
-    shim_folder(path, path.rstrip("/").split("/")[-1], shim_path)
+    edited = shim(path, shim_dir)
+    return edited or None
```
```suggestion
    return edited
```

onerandomusername replied:
This is intentional because I don't want to return an empty list.

shiftinv commented on shim_maker.py:
```diff
@@ -21,62 +26,138 @@ def fancy_list(prefix: str, to_join: Iterable[str], limit: int = 100, with_brack
+        print(len(res), "files modified.")
+    else:
+        print("No changes were made.")
+    sys.exit(bool(res))
```
```suggestion
```

shiftinv reviewed (changes requested):
pretty cool stuff, but some type hints are a bit incorrect and the file reading/writing could be simplified since it's using pathlib already

onerandomusername committed:
shim: run generation

onerandomusername committed:
chore: run isort over all files and rewrite list exporter

onerandomusername committed:
ci: add precommit

onerandomusername committed:
chore: refactor read/write logic

onerandomusername committed:
chore: rerun shim generation

onerandomusername committed:
fix: ensure that runtime loaded modules are imported as well

onerandomusername committed:
chore: don't import from inits

onerandomusername committed:
feat: add a test shim script

onerandomusername committed:
fix: overwrite versioninfo as well

onerandomusername committed:
tester: more verboseness

onerandomusername committed:
chore: generate shim

shiftinv commented on test_shim.py:
```diff
@@ -0,0 +1,67 @@
+        if attr.startswith("__"):
+            continue
+        base_attr: object = getattr(base, attr)
+        if not getattr(base_attr, "__package__", None) == base_name.split(".", 1)[
```
`base_name.split(".", 1)[0]` could be outside of the loop

shiftinv replied:
also, would be nice if there were some comments explaining what these two conditions are actually doing :)

onerandomusername replied:
As per https://github.com/DisnakeDev/discord-disnake/pull/3/commits/95851206b78d4230b7e95137aa763108a916c395 these should now be commented!

shiftinv commented on test_shim.py:
```diff
@@ -0,0 +1,67 @@
+
+def test_shim(
+    base: ModuleType, shim: ModuleType, base_name: str, _recurse_modules: bool = True
+) -> List[pathlib.Path]:
```
method doesn't return any paths
(should it?)

onerandomusername replied:
Now returns an int, and changed, thanks for the eye (and private suggestion to set pyright to basic typechecking!)

shiftinv commented on test_shim.py:
```diff
@@ -0,0 +1,67 @@
+            # print(attr, " is the same")
+
+
+def main() -> int:
```
doesn't return an int

onerandomusername replied:
it does now, thanks for noticing

onerandomusername committed:
chore: add timers

onerandomusername committed:
minor: address review

onerandomusername committed:
fix: accurate typehints in test_shim

shiftinv commented on shim_maker.py:
```diff
@@ -1,45 +1,40 @@
-        return prefix + "(\n    " + as_line + "\n)"
 
-    return prefix + "(\n    " + ",\n    ".join(to_join) + ",\n)"
+def fancy_list(prefix: str, to_join: Iterable[str]) -> str:
```
`len` doesn't work on `Iterable`s, changing it to `List[str]` won't work either since it could also be a set when called from `shim_init`. I suppose that one could be changed to a list instead, or use `Collection` instead of `Iterable` and change the `to_join[0]` part

onerandomusername replied:
`Sequence` will work here

shiftinv commented on shim_maker.py:
```diff
@@ -53,56 +48,193 @@ def create_file(module_name: str) -> Tuple[str, str]:
+    # However, because they are lazily loaded, we don't want them showing up on intellisense
+    # so should not be reaching when typechecking.
+    from typing import TYPE_CHECKING
+    """
```
```suggestion
        """
        # Because the main library lazy loads some files, its important to re-export them here.
        # However, because they are lazily loaded, we don't want them showing up on intellisense
        # so should not be reaching when typechecking.
        from typing import TYPE_CHECKING
        """
```

shiftinv commented on shim_maker.py:
```diff
@@ -53,56 +48,193 @@ def create_file(module_name: str) -> Tuple[str, str]:
-    os.makedirs("discord", exist_ok=True)
-    shim_folder(path, path.rstrip("/").split("/")[-1], shim_path)
+if __name__ == "__main__":
+    import sys
+    import time
```
these should be moved to the top of the file

onerandomusername replied:
They intentionally aren't because no where else uses them, so if used by something else programmably they don't need to be imported.

shiftinv replied:
Fair enough, in that case `sys` is still being imported twice though

shiftinv commented on shim_maker.py:
```diff
@@ -53,56 +48,193 @@ def create_file(module_name: str) -> Tuple[str, str]:
+        except ImportError:
+            print(f"Invalid path to package to be shimmed: {args.path}")
+            return 2, []
+        path = pathlib.Path(mod.__file__).parent
```
```suggestion
        path = pathlib.Path(mod.__file__).parent  # type: ignore
```

shiftinv commented on shim_maker.py:
```diff
@@ -53,56 +48,193 @@ def create_file(module_name: str) -> Tuple[str, str]:
+    return res
+
+
+def main() -> Tuple[int, list]:
```
```suggestion
def main() -> Tuple[int, List[pathlib.Path]]:
```

shiftinv commented on shim_maker.py:
```diff
@@ -53,56 +48,193 @@ def create_file(module_name: str) -> Tuple[str, str]:
+    code += "\n\n"
+    code += "# doubly ensure that everything is overwritten. Most of the above exist just for typechecking."
+    # overwrite the VersionInfo and version_info to be from the original.
+    code += f"# isort: split\nfrom {path.stem} import *"
```
should probably do the `__original_dict__` thing here again, pyright doesn't seem to like this import (`Wildcard import from a library not allowed`)

onerandomusername replied:
Added in https://github.com/DisnakeDev/discord-disnake/pull/3/commits/62a3051fb1222b91f34049399a6c3940d03a4f18, however, this broke the shim.

Consequentially, removed in https://github.com/DisnakeDev/discord-disnake/pull/3/commits/e5b7218b43021785add4b89d666cd9e512aa22ce for reasons explained in the comment and commit message.

shiftinv reviewed (commented):
looks good, got a few more tiny issues; also:
- could you run isort/black on the scripts? otherwise pre-commit will hate it when this gets moved to the main repo again
- what about CRLF/LF differences between operating systems? if the script just changes the newlines (because git converted them to CRLF or something like that), there would be tons of unnecessary file operations (and pre-commit would fail, if we decide to use that for shim generation)
- the `opus.py` thing mentioned above is still an issue, not sure what to do about that... maybe just make an exception for those, if there's no better way to handle it (?)

onerandomusername committed:
chore: type comments, formatting

Co-authored-by: shiftinv <8530778+shiftinv@users.noreply.github.com>

onerandomusername committed:
chore: delete files that were deleted on master

this is done by removing the existing shim and generating an all new shim every run
This can be optimized, but a full run completes in half a second right now
so its not worth the effort to me at this time.

onerandomusername committed:
shim: run regen

onerandomusername committed:
chore: remove debugging print statement

onerandomusername committed:
fix: make the extension shims work again

reverts updating the dictionary in __init__.py
that somehow broke the shims and given it works without that
I don't want to figure out why it breaks it

onerandomusername committed:
shim: run regen

onerandomusername committed:
tests(shim): test all module files exist and no extra files exist

onerandomusername committed:
chore: catch loading errors and get straight to the point

if a shim file fails to load, stop testing then

onerandomusername committed:
feat: rewrite the shim tester from scratch

onerandomusername commented:
I completely rewrote the shim tester from scratch to make it easier to read and understand.

It tests that all modules exist and the attributes within them are the same.

onerandomusername committed:
minor: use coloredlogs for logging if its installed

onerandomusername committed:
chore: optimize find modules to work even in a different working directory

onerandomusername committed:
feat: rewrite the shim generator

onerandomusername commented:
Don't like the existing generator, I'm going to fully rewrite it.

onerandomusername committed:
fix: don't update private attributes

onerandomusername committed:
chore(test): report mismatched attributes

onerandomusername committed:
chore: format __all__ properly

onerandomusername committed:
fix: actually find missing attributes with the shim tester

onerandomusername committed:
fix: keep existing __module__ attribute

onerandomusername committed:
chore: add test task to taskipy

onerandomusername committed:
fix: check that the base attribute is defined in that file

onerandomusername committed:
feat: make the shim a 1:1 perfect clone

onerandomusername committed:
fix: special handle __init__ files for typechecking

onerandomusername committed:
fix: make __main__ runnable

onerandomusername committed:
chore: only MainLoader if fullname ends with main

onerandomusername committed:
fix: properly handle ext.commands and __main__

onerandomusername committed:
fix: TEMP for setup.py to work with no version string

onerandomusername committed:
fix: weird case of the wrong file being imported

onerandomusername committed:
fix: support third party packages in the discord namespace

onerandomusername committed:
fix: support third party modules when imported absolutely

eg import discord.ext.menus

onerandomusername committed:
fix: keep modules referencing disnake and not discord

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+
+class MainLoader(importlib.abc.ExecutionLoader):
+    def get_filename(self, fullname):
+
```
nitpick: Remove this empty line

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+
+        spec = importlib.util.find_spec(fullname)
+        if not spec or not spec.origin:
+            raise ImportError
```
Shouldn't this provide more info, given it can be called for multiple files/ocassions?

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+        if file := self.get_filename(fullname):
+            with open(file, "r") as f:
+                return f.read()
+        raise ImportError
```
See above

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+            return f.read()
+
+    def get_filename(self, fullname):
+
```
nitpick: Remove empty line

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+        if file := self.get_filename(fullname):
+            with open(file, "r") as f:
+                return f.read()
+        raise ImportError
```
See earlier comment about details

onerandomusername replied:
Indeed, that line of code is unreachable.

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+        # Get the path to the module
+        path = self.get_filename(module.__spec__.name)
+        if not path:
+            raise RuntimeError("Could not find module")
```
Add the module you couldn't find into this error string?

onerandomusername replied:
should also be an import error too.

To address your other ImportError comments, most of them should never ever ever ever run, if we're lucky.

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+        # force the module to be an attribute
+        if "." not in module.__spec__.name:
+            return
+        shimparent, name = module.__spec__.name.rsplit(".", 1)
```
nitpick: ``shim_parent``

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on discord/__init__.py:
```diff
@@ -1,77 +1,121 @@
+            if parent not in sys.modules:
+                importlib.import_module(parent)
+            # extremely hacky
+            ogparentmodule = sys.modules[parent]
```
nitpick: ``og_parent_module``

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/0df9eccf75dab4d569bbb1878b1cfacabc8e8c7c

Skelmis commented on setup.py:
```diff
@@ -1,41 +1,38 @@
 
-if not version:
-    raise RuntimeError("version is not set")
+VERSION = "2.3.0"
```
Is this hard-coded for a reason?

onerandomusername replied:
Yes, once this works we will be moving it to the main repo where it will be able to use the disnake version.

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+
+def find_packages(dir: str, package: str = None) -> List[str]:
+    """
+    Finds all of the submodules in the designated modules and returns a list of them.
```
Document arguments

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+
+
+def _type_shim_code(code: ModuleType, is_init=False) -> str:
+    """Provided code, returns a type-stub version of it."""
```
nitpick cos private: Document arguments

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+    base_module: ModuleType, shim_module: str, *, original_shim: pathlib.Path = None
+):
+    """
+    Shim a module into the other location.
```
nitpick cos private: document arguments

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+    Shim a module into the other location.
+    """
+    if not base_module.__file__:
+        raise RuntimeError(f"{base_module} does not have a __file__ attribute")
```
If this error is to be believed, would this not throw an ``AttributeError`` rather then play nice in the if?

onerandomusername replied:
good catch

onerandomusername commented:
Starting with https://github.com/DisnakeDev/discord-disnake/pull/3/commits/6e6ac68dc8d9702f3c66662cd1614412aae862ad the shim uses importlib, and pep 302 import hooks in order to make the modules the same module. This means that `import discord` **will return the disnake module**. This means that a tester and shim generator are *mostly* useless. 

The v4 shim generator (as written in https://github.com/DisnakeDev/discord-disnake/pull/3/commits/286bf5c488e1da51babad6875a2c036088f1fbb9) has been modified to export public attributes, and public attributes only, as its sole purpose is to provide `.pyi` files.


In total, this pr has shim generator v2 to v5, summarized below

- v2 https://github.com/DisnakeDev/discord-disnake/pull/3/commits/3ca4c75cfb43e37db2182e20de1b84a34db1546d to https://github.com/DisnakeDev/discord-disnake/pull/3/commits/733fff73d8a98fb85356e898d0e5a1ac72f93cef the initial "rewrite", added quite a few support options but had some bugs
- v3 https://github.com/DisnakeDev/discord-disnake/pull/3/commits/286bf5c488e1da51babad6875a2c036088f1fbb9 to https://github.com/DisnakeDev/discord-disnake/pull/3/commits/6c440a4f999eb81dceed20119fa233c817cf7ccf a fresh shim generator, written from scratch, with the end goal the same v2. Later modified to provide typing stubs for v4 and v5
- v4 and v5 https://github.com/DisnakeDev/discord-disnake/pull/3/commits/6e6ac68dc8d9702f3c66662cd1614412aae862ad to https://github.com/DisnakeDev/discord-disnake/pull/3/commits/c15e8086f86dc7635ff3dc0ec04e8c96c9a88ccc and beyond. Uses pep 302 import hooks in order to fully shim the entire module without leaving a weird import mess. This ensures that disnake is discord. This version is entirely version agnostic, and does not depend on a specific version of disnake, except for typing stubs, which have no impact at runtime.

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+
+
+def shim_module(base_name, shim_name, module_name, original_shim: pathlib.Path = None):
+    """Shim a module and all of its submodules."""
```
Document arguments, also possible types if applicable?

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+
+
+def main(base_name: str, shim_name: str):
+
```
nitpick: remove empty line

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+
+    base = importlib.import_module(base_name)
+    if not base.__file__:
+        raise RuntimeError(f"{base_name} does not have a __file__ attribute")
```
See earlier comment about this

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+        raise RuntimeError(f"{base_name} does not have a __file__ attribute")
+    base_dir = os.path.dirname(base.__file__)
+    # i don't care enough at the moment to do this the right way
+    original_init = shim_name + os.sep + "__init__.py"
```
Is this not just `os.path.join` though?

Skelmis commented on test_shim.py:
```diff
@@ -0,0 +1,225 @@
+        # handle main.
+        # main is special, since it needs to have equal attributes, but not the same attributes
+        # we can get a near perfect result for this by comparing that the signatures, docstrings, and so forth are equal.
+        # TODO:
```
Empty todo, or misplaced todo?

onerandomusername replied:
Todo is for the above. Additionally, will be killing this tester too, as it doesn't properly test the current iteration of the shim.

Skelmis commented on shim_maker.py:
```diff
@@ -1,108 +1,156 @@
+        )
+
+
+def main(base_name: str, shim_name: str):
```
Doesnt this need to return an ``int`` for ``sys.exit``

onerandomusername replied:
Per [the docs](https://docs.python.org/3/library/sys.html#sys.exit)
> If another type of object is passed, None is equivalent to passing zero, and any other object is printed to stderr and results in an exit code of 1

Skelmis commented on test_shim.py:
```diff
@@ -0,0 +1,225 @@
+    start = time.time_ns()
+    res = main()
+    end = time.time_ns()
+    print(f"Took {(end - start) / 1e9} seconds.")
```
Shouldn't this use logging?

onerandomusername replied:
IMO, no. I want to be able to redirect stderr and keep this in the console.

onerandomusername committed:
chore: address review, remove empty lines, raise proper exceptions

onerandomusername committed:
chore: delete shim tester, its no longer used nor works properly

onerandomusername committed:
chore: minor fixes

shiftinv commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+generate = { cmd= "python shim_maker.py", help = "Generate a new version of the shim" }
+lint = { cmd = "pre-commit run --all-files", help = "Checks all files for linting errors" }
+precommit = { cmd = "pre-commit install --install-hooks", help = "Installs the precommit hook" }
+test = { cmd = "python test_shim.py disnake discord", help = "Test the shim functions properly"}
```
```suggestion
```

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+ISORT_CONFIG = pathlib.Path("pyproject.toml")
+
+
+def find_packages(dir: str, package: str = None) -> List[str]:
```
```suggestion
def find_packages(dir: str, package: Optional[str] = None) -> List[str]:
```

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
-        public_members = None
+import textwrap
+from types import ModuleType
+from typing import List
```
```suggestion
from typing import List, Optional
```

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    return sorted([submod[:-9] for submod in submodules if submod.endswith("__init__")])
+
+
+def _type_shim_code(code: ModuleType, is_init=False) -> str:
```
```suggestion
def _type_shim_code(code: ModuleType, is_init: bool = False) -> str:
```

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    if hasattr(code, "__all__") and not is_init:
+        imports.update(code.__all__)
+
+        if len(code.__all__) > 1:
```
Why bother with formatting this manually instead of running black over the resulting code?

onerandomusername replied:
Don't want to depend on black, and it doesn't have a stable api

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    return shim.strip() + "\n"
+
+
+def _shim_module_type(
+    base_module: ModuleType, shim_module: str, *, original_shim: pathlib.Path = None
+):
```
```suggestion
def _shim_module_type(
    base_module: ModuleType, shim_module: str
) -> None:
```
all the other `original_shim` code below is unused as well

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+        shim_path += ".pyi"
+        is_init = False
+
+    if not os.path.exists(os.path.dirname(shim_path)):
+        os.makedirs(os.path.dirname(shim_path))
```
```suggestion
    os.makedirs(os.path.dirname(shim_path), exist_ok=True)
```

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+        )
+
+
+def main(base_name: str, shim_name: str):
```
```suggestion
def main(base_name: str, shim_name: str) -> None:
```
or alternatively `Any`

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    for package in packages:
+        shim_module(base_name, shim_name, package)
+
+    with open(original_init, "w") as f:
```
all the `open()` calls in this code should use `encoding="utf8"`, as it's not the default on Windows systems

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+        """
+        Executes the module.
+        """
+        # Get the path to the module
```
```suggestion
        if not module.__spec__:
            return
        # Get the path to the module
```
typecheckers don't like the fact that `__spec__` can be `None`, not sure if that's possible here though

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/a7cbb406025e5a5ae96eb4eb12591da5c03103ff

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    """
+    submodules = set()
+    package = package or dir
+    for path in glob.iglob(f"{dir}/**/*.py", recursive=True):
```
```suggestion
    for path in glob.iglob(f"{dir}/**/__init__.py", recursive=True):
```

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    """
+    submodules = set()
+    package = package or dir
+    for path in glob.iglob(f"{dir}/**/*.py", recursive=True):
+        path = path[len(dir) : -3].replace(os.sep, ".")
+        if importlib.util.find_spec(path, package) is not None:
+            submodules.add(path)
+
+    return sorted([submod[:-9] for submod in submodules if submod.endswith("__init__")])
```
```suggestion
    for path in glob.iglob(f"{dir}/**/__init__.py", recursive=True):
        path = path[len(dir) : -3].replace(os.sep, ".")
        if importlib.util.find_spec(path, package) is not None:
            submodules.add(path[:-9])

    return sorted(submodules)
```

onerandomusername replied:
This feels too obvious...

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+    with open(shim_path, "w") as f:
+        f.write(code)
+
+    if os.path.exists(py_typed := os.path.dirname(base_module.__file__) + os.sep + "py.typed"):
+        shutil.copyfile(py_typed, os.path.dirname(shim_path) + os.sep + "py.typed")
```
couldn't this be moved into `main`?

onerandomusername replied:
No, because its *possible* for a sub package to have py.typed afaict

ooliver1 replied:
For namespace packages yes

shiftinv commented on shim_maker.py:
```diff
@@ -1,108 +1,162 @@
+        raise RuntimeError(f"{base_name}'s __file__ attribute is not set")
+    base_dir = os.path.dirname(base.__file__)
+    # i don't care enough at the moment to do this the right way
+    original_init = os.path.join(shim_name, "__init__.py")
```
I don't see what this part of the code is useful for

onerandomusername replied:
its no longer used, but its from when this generated .py files instead of .pyi

onerandomusername replied:
Its now used.

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/951fd6e4b2d574238926f123d68b9231ac8f225f

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+        raise ModuleNotFoundError("No module named '{}'".format(fullname))
+
+
+class Loader(importlib.abc.SourceLoader, importlib.abc.Loader):
```
```suggestion
class Loader(importlib.abc.SourceLoader):
```

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/a7cbb406025e5a5ae96eb4eb12591da5c03103ff

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+
+    def create_module(self, spec):
+        return None
+
```
```suggestion
```
this is just the default implementation, as far as I can tell

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/a7cbb406025e5a5ae96eb4eb12591da5c03103ff

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+MAIN_PACKAGE = "disnake"
+
+
+def _find_file(fullname, path=None):
```
why can't this be replaced with `importlib.util.find_spec`?

onerandomusername replied:
because this is used in finding said spec.

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+        return spec.origin
+
+    def get_source(self, fullname):
+        if file := self.get_filename(fullname):
+            with open(file, "r") as f:
+                return f.read()
+        raise ModuleNotFoundError("No module named '{}'".format(fullname))
```
```suggestion
        with open(self.get_filename(fullname), "r") as f:
            return f.read()
```
`get_filename` can't return `None`

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/a7cbb406025e5a5ae96eb4eb12591da5c03103ff

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+
+class Loader(importlib.abc.SourceLoader, importlib.abc.Loader):
+    def get_data(self, path):
+        with open(path, "r") as f:
```
```suggestion
        with open(path, "rb") as f:
```
should return bytes

onerandomusername replied:
https://github.com/DisnakeDev/discord-disnake/pull/3/commits/a7cbb406025e5a5ae96eb4eb12591da5c03103ff

shiftinv commented on discord/__init__.py:
```diff
@@ -1,77 +1,117 @@
+        return _find_file(fullname)
+
+    def get_source(self, fullname):
+        with open(self.get_filename(fullname), "r") as f:
```
not sure if this should use the default encoding; fwiw, the builtin loaded tries to detect the encoding based on the file contents

onerandomusername committed:
address some reviews

onerandomusername committed:
Merge branch 'master' into rewrite-shim

onerandomusername committed:
chore: run lint

onerandomusername committed:
chore: add encoding to all open calls

onerandomusername committed:
address some reviews

onerandomusername committed:
minor: update readme

onerandomusername committed:
add package blacklist

shiftinv commented:
This would've been quite cool, but given that we're archiving the repo, no longer necessary. ty :pray:
